### PR TITLE
[fix][broker] Reader stuck after call hasMessageAvailable when enable replicateSubscriptionState

### PR DESCRIPTION
**Pull request by shibd**
### Motivation

#22571

### Analysis

When enabling `replicateSubscriptionState` will use topic to sync subscription state, and make these message metadata as [Marker](https://github.com/apache/pulsar/blob/ac94296d2488b2b13d76fa2b1bfa71e9e3cfbb45/pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Markers.java#L107-L120)

These `marker` messages will not be sent to the consumer by the topic, and will automatically ack them.

https://github.com/apache/pulsar/blob/fc393f69043be6eb1b2572a27f131656a2cbc7f6/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java#L202-L217

But `getLastMessageId` will always return the last message position, regardless of whether the last message is `marked` or `not`. This will cause the reader stuck.

```java
        while (reader.hasMessageAvailable()) {  // get true
              Message message reader.readNext();  // never can't receive msg.
        }
```

You can refer to this diagram to help understand this bug:

image

### Modifications
- Add `asyncReverseFindPositionOneByOne` method on `ManagedLedger` .
- Add `getLastCanDispatchPosition` method on `Topic`, it will call `asyncReverseFindPositionOneByOne` to find the last position of entry that not is `replistateSubscriptionState` 
- Change the `getLastMessageId` implement of `ServerCnx` to use `getLastCanDispatchPosition` instead of `getMaxReadPosition`.

### Verifying this change
- Add `ManagedLedgerTest.testReverseFindPositionOneByOne` to cover ReverseFindPositionOneByOne method.
- Add `testReplicatedSubscriptionAcrossTwoRegionsGetLastMessage` to cover this bug.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java [hunk 1]
```diff
@@ -4321,4 +4321,44 @@ public void testDeleteCurrentLedgerWhenItIsClosed(boolean closeLedgerByAddEntry)
...
+        String noMatchEntry = "nomatch-entry";
+        Predicate<Entry> predicate = entry -> {
+            try {
+                String entryValue = entry.getDataBuffer().toString(UTF_8);
+                if (matchEntry.equals(entryValue)) {
+                    return true;
+                }
+            } finally {
+                entry.release();
+            }
```
Comment (lhotari) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java:4343]: it's better to move releasing to internalAsyncReverseFindPositionOneByOne ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 2]
```diff
@@ -509,6 +509,11 @@ public PositionImpl getMaxReadPosition() {
         }
     }
 
+    @Override
+    public CompletableFuture<Position> getLastCanDispatchPosition() {
+        return CompletableFuture.completedFuture(getMaxReadPosition());
```
Comment (dao-jun) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:514]: It looks a little strange, after recover finished and there is no ongoing txns, the maxReadPosition will be set to ledger's lastConfirmedEntry, and the last entry also could be a Marker.

Comment (dao-jun) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:514]: And if all the txns committed/aborted, the maxReadPosition will also set to lastConfirmedEntry, do we need to consider the case?

Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:514]: I agree with your point, I'm not familiar with transaction implementation yet, and I'm not sure if there's anything special to deal with. 

This PR will keep the logic the same as before when enabling transactions. and I think I can write a test in the next PR to verify your guess and fix the transaction buffer issue.

Comment (dao-jun) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:514]: I mean if we want to fix #22571 by adding `getLastCanDispatchPosition` here, we also need to consider the `maxReadPosition` equals to `lastConfirmedEntry` and it could be a Marker.

Before 
```java
return CompletableFuture.completedFuture(getMaxReadPosition());
```
verify the `maxReadPosition` is valid.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java (hunk 1 above)
Comment (shibd) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java]: Make sense, moved.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 2 above)
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java]: Make sense to me, I add this logic and add `ReplicatorSubscriptionWithTransactionTest` to cover treansaction enable case. Thanks for your reviews.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBuffer.java [hunk 3]
```diff
@@ -160,6 +160,12 @@ public interface TransactionBuffer {
      */
     PositionImpl getMaxReadPosition();
 
+    /**
+     * Get the can dispatch max position.
+     * @return the stable position.
+     */
+    CompletableFuture<Position> getLastCanDispatchPosition();
```
Comment (dao-jun) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBuffer.java:167]: Maybe we don't need this method, and move `TopicTransactionBuffer#getLastCanDispatchPosition`'s logic to `PersistentTopic#getLastCanDispatchPositionWithTxn()` is better.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 4]
```diff
@@ -1852,6 +1852,51 @@ public void findEntryFailed(ManagedLedgerException exception, Optional<Position>
         return future;
     }
 
+    @Override
+    public CompletableFuture<Position> asyncReverseFindPositionOneByOne(Predicate<Entry> predicate) {
```
Comment (dao-jun) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: Maybe `asyncFindLastValidPosition` or something else is better

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBuffer.java (hunk 3 above)
Comment (dao-jun) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBuffer.java:167]: Maybe we don't need this method in TransactionBuffer, and move `TopicTransactionBuffer#getLastCanDispatchPosition`'s logic to `PersistentTopic#getLastCanDispatchPositionWithTxn()` is better.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 5]
```diff
@@ -3602,6 +3603,15 @@ public Position getLastPosition() {
         return ledger.getLastConfirmedEntry();
     }
 
+    @Override
+    public CompletableFuture<Position> getLastCanDispatchPosition() {
```
Comment (dao-jun) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3607]: Maybe we can make the method private or move the logics to `getLastCanDispatchPositionWithTxn()`, just expose 
`getLastCanDispatchPositionWithTxn()`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 6]
```diff
@@ -4082,6 +4092,10 @@ public PositionImpl getMaxReadPosition() {
         return this.transactionBuffer.getMaxReadPosition();
     }
 
+    public CompletableFuture<Position> getLastCanDispatchPositionWithTxn() {
```
Comment (dao-jun) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:4095]: Maybe we can rename the method to `getLastDispatchablePositionWithTxn` or something else.

**Review by dao-jun — APPROVED** [review]
Overall looks good to me, just left some minor comments about the code style

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 4 above)
Comment (shibd) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: I want to show two key points in the method name: `Reverse` and `OneByOne`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 7]
```diff
@@ -509,6 +509,16 @@ public PositionImpl getMaxReadPosition() {
         }
     }
 
+    @Override
+    public CompletableFuture<Position> getLastCanDispatchPosition() {
+        PositionImpl tnxMaxReadPosition = getMaxReadPosition();
+        if (topic.getLastPosition() == tnxMaxReadPosition) {
```
Comment (dao-jun) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:515]: I know `tnxMaxReadPosition` and `topic.getLastPosition()` are the same object if they are equals to, but for the compare about two `PositionImpl` objects, `topic.getLastPosition.compareTo(tnxMaxReadPosition) == 0` is better

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 5 above)
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3607]: About the simple method comment, I know your intention.

After these change, there will be fewer methods. But this logic runs into the `persistentTopic`.
```java
        PositionImpl tnxMaxReadPosition = transactionBuffer.getMaxReadPosition();
        if (getLastPosition() == tnxMaxReadPosition) {
            return topic.getLastCanDispatchPosition();
        } else {
            return CompletableFuture.completedFuture(tnxMaxReadPosition);
        }
```

I still want it to be cohesive in class `TransactionBuffer`.

I think makes sense to expose the `getLastCanDispatchPosition` method in `TransactionBuffer` and `Topic`, it can clearly remind the caller:  **The position returned by the `getLastPosition` method is not dispatchable, should use `getLastCanDispatchPosition`**

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 6 above)
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:4095]: Make sense to me.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 5 above)
Comment (dao-jun) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: For TransactionBuffer, it just handle the TXN commit/abort/append requests and provide the guarantee of READ_COMMIT.
Determine the position to dispatch is Topic's duty, TransactionBuffer should not be aware of it.
See also https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L4122-L4124

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 4 above)
Comment (BewareMyPower) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: I also support the `asyncFindLastValidPosition` name. `Reverse` and `OneByOne` just describe how do you achieve the goal, but `findLastValidPosition` describes what goal you want to achieve.

If the way of how to achieve the goal changes in future, the method name might not be proper.

Comment (RobertIndie) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: +1 for `asyncFindLastValidPosition`. We are defining the interface method instead of the implementation. `asyncReverseFindPositionOneByOne` is to reflect the implementation details while `asyncFindLastValidPosition` is more suitable for the interface.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java [hunk 8]
```diff
@@ -648,6 +648,11 @@ default void skipNonRecoverableLedger(long ledgerId){}
      * */
     CompletableFuture<Position> asyncFindPosition(Predicate<Entry> predicate);
 
+    /**
+     * Reverse find position one-entry by one-entry.
+     */
+    CompletableFuture<Position> asyncReverseFindPositionOneByOne(Predicate<Entry> predicate);
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java:654]: The ManageLedger has been marked as `@InterfaceStability.Stable`. So we need to have a default implementation for the newly added method.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 4 above)
Comment (codelipenghui) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: We already have `findNewestMatching` method in the ManageCursor API. We can just use the ReadonlyCursor for find the the available position. So that we don't need to make any changes to the ManagedLedger API.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Topic.java [hunk 9]
```diff
@@ -275,6 +275,11 @@ CompletableFuture<? extends TopicStatsImpl> asyncGetStats(boolean getPreciseBack
 
     Position getLastPosition();
 
+    /**
+     * Get the last message position that can be dispatch.
+     */
+    CompletableFuture<Position> getLastDispatchablePosition();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Topic.java:281]: It's also better to have a default implementation.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 10]
```diff
@@ -509,6 +509,16 @@ public PositionImpl getMaxReadPosition() {
...
 
+    @Override
+    public CompletableFuture<Position> getLastDispatchablePosition() {
+        PositionImpl tnxMaxReadPosition = getMaxReadPosition();
+        if (tnxMaxReadPosition.compareTo((PositionImpl) topic.getLastPosition()) == 0) {
+            return topic.getLastDispatchablePosition();
+        } else {
+            return CompletableFuture.completedFuture(tnxMaxReadPosition);
+        }
+    }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:520]: If we want to introduce this method in the TransactionBuffer. I think we'd better make sure the behavior is correct. The aborted transactions is not dispatchable, so we'd filter them out.

Or we can just revert this change and open another PR to handle the transaction issues which will be more better approach because the PR title said it fixes the issue related to replicate subscription state.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java [hunk 11]
```diff
@@ -167,6 +168,82 @@ public void testReplicatedSubscriptionAcrossTwoRegions() throws Exception {
...
+        try (Consumer<byte[]> consumer1 = client1.newConsumer()
+                .topic(topicName)
+                .subscriptionName(subscriptionName)
+                .replicateSubscriptionState(replicateSubscriptionState)
+                .subscribe()) {
+            readMessages(consumer1, receivedMessages, 3, false);
+        }
+
+        // wait for subscription to be replicated
+        Thread.sleep(2 * config1.getReplicatedSubscriptionsSnapshotFrequencyMillis());
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:230]: Is it possible to avoid the Thread.sleep by employ the Awaitability?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 4 above)
Comment (BewareMyPower) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: It might not be worth creating another cursor.

I think we can use add a static method because all the methods used are public or package visiable.

```java
@InterfaceStability.Evolving
public class ManagedLedgerImplUtils {

    public static CompletableFuture<Position> asyncGetLastValidPosition(final ManagedLedgerImpl ledger,
                                                                        final Predicate<Entry> predicate) {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 10 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:520]: +1

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 12]
```diff
@@ -3625,6 +3626,15 @@ public Position getLastPosition() {
...
     }
 
+    @Override
+    public CompletableFuture<Position> getLastDispatchablePosition() {
+        return ledger.asyncReverseFindPositionOneByOne((entry -> {
+            MessageMetadata md = Commands.parseMessageMetadata(entry.getDataBuffer());
+            // If a messages has marker will filter by AbstractBaseDispatcher.filterEntriesForConsumer
+            return !Markers.isServerOnlyMarker(md);
+        }));
+    }
```
Comment (coderzc) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3636]: It is better to return a smaller value with `getMaxReadPosition()`.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 4 above)
Comment (shibd) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: > We already have findNewestMatching method in the ManageCursor API. We can just use the ReadonlyCursor for find the the available position. So that we don't need to make any changes to the ManagedLedger API.

@codelipenghui Here need `Reverse` find, and this method doesn't seem match.

@BewareMyPower @RobertIndie @tjiuming What do you think call `asyncReverseFindLastValidPosition`?  I think `Reverse` should be part of the method name.


> I think we can use add a static method because all the methods used are public or package visible.

Is there any harm to adding an interface? I'd prefer to embody it on the interface, which shows the various use cases of Pulsar for the `ManangerLedger`.

Comment (BewareMyPower) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: > Is there any harm to adding an interface?

See https://github.com/apache/pulsar/pull/22572#discussion_r1582003060

```java
  /**
   * Can evolve while retaining compatibility for minor release boundaries.;
   * can break compatibility only at major release (ie. at m.0).
   */
```

So you have to use a default method or a static method.

> What do you think call asyncReverseFindLastValidPosition? I think Reverse should be part of the method name.

`Reverse` is still redundant. You can take the Java standard API for example:

```java
        String s = "hello";
        s.lastIndexOf('l');
```

It's `lastIndexOf`, no something like `reverse`. Because `Last` is already a hint for reverse finding. Python string API has `index` and `rindex`, where `r` is the hint for `reverse`. But it does not have any other hint like `last`.

Comment (BewareMyPower) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: Regarding implementation, actually we can use sequential find on the active ledger rather than the reverse finding because we already have the cache. We only need reverse finding when all entries in the active ledger are not available.

So "Reverse" is not required.

Comment (shibd) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: Thanks for explaining, I will accept this suggestion.

https://github.com/apache/pulsar/pull/22572#discussion_r1582006268

And, I will keep `reverse ` find, Because in most cases, the last entity can meet the predicate.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 12 above)
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3636]: Why, since IO may be involved here, it is more appropriate to return to the future

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 4 above)
Comment (codelipenghui) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1856]: > @codelipenghui Here need Reverse find, and this method doesn't seem match.

Oh I see. Thanks for the explanation. A new method is good to me.

And I think we can follow @BewareMyPower's suggestion to add a Until class. 

> Regarding implementation, actually we can use sequential find on the active ledger rather than the reverse finding because we already have the cache. We only need reverse finding when all entries in the active ledger are not available.

I didn't get the point. Even if the entries are cached, reverse reading is also more efficient for getting the last valid entry?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java (hunk 11 above)
Comment (shibd) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:230]: I didn't find a valid condition to determine that the subscription status is synced. 

This test class uses the same sleep code in a lot of places, it looks like it's stable, maybe we can keep using it?

**Review by codelipenghui — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 4 above)
Comment (BewareMyPower) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: It's just an example to show the implementation could vary. Not a suggestion to the actual implementation. I meant
1. `read(activeLedger)` returns a list of entries (`entries`)
2. `for (int i = entries.size() - 1; i >= 0; i--)` reverse find

not

```
while (true) {
  var position = getPreviousPosition(position);
  var entry = readOneEntry(position);
  /* ... */
}
```

The code above is pseudo code. When I mentioned the implementation, I didn't look into the details of the cache. So please don't take serious on my example.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java [hunk 13]
```diff
@@ -167,6 +168,82 @@ public void testReplicatedSubscriptionAcrossTwoRegions() throws Exception {
...
+        // create a reader in r2
+        Reader<byte[]> reader = client2.newReader().topic(topicName)
+                .subscriptionName("new-sub")
+                .startMessageId(MessageId.earliest)
+                .create();
+        int readNum = 0;
+        while (reader.hasMessageAvailable()) {
+            Message<byte[]> message = reader.readNext(10, TimeUnit.SECONDS);
+            log.info("Receive message: " + new String(message.getValue()) + " msgId: " + message.getMessageId());
+            assertNotNull(message);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:241]:
```suggestion
            assertNotNull(message);
            log.info("Receive message: " + new String(message.getValue()) + " msgId: " + message.getMessageId());
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 12 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Oh, sorry, I mean if the result of `asyncReverseFindPositionOneByOne` is greater than `getMaxReadPosition`, then the result of `getMaxReadPosition` should be returned.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 14]
```diff
@@ -3634,6 +3636,22 @@ public Position getLastPosition() {
...
+        if (maxReadPosition.compareTo((PositionImpl) getLastPosition()) != 0) {
+            return CompletableFuture.completedFuture(maxReadPosition);
+        } else {
+            return ManagedLedgerImplUtils.asyncGetLastValidPosition((ManagedLedgerImpl) ledger, entry -> {
+                MessageMetadata md = Commands.parseMessageMetadata(entry.getDataBuffer());
+                // If a messages has marker will filter by AbstractBaseDispatcher.filterEntriesForConsumer
+                return !Markers.isServerOnlyMarker(md);
+            });
+        }
+    }
```
Comment (coderzc) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3653]: Not sure if there is a race condition here, if produce uncommitted transactions before searching a valid position, then we can return an uncommitted position.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 12 above)
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: > Oh, sorry, I mean if the result of asyncReverseFindPositionOneByOne is greater than getMaxReadPosition, then the result of getMaxReadPosition should be returned.

I  think the result of `asyncReverseFindPositionOneByOne` is never greater than `getMaxReadPosition()`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 14 above)
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3653]: +1, we need to pass `maxReadPosition` to `asyncGetLastValidPosition`, thanks for your reminder.

**Review by coderzc — APPROVED** [review]
LGTM

**Comment by dao-jun** [comment]
The PR handled the case of ServerOnlyMarker, but it looks we also need to handle txn aborted messages.
I created a PR for improvement, PTAL
https://github.com/apache/pulsar/pull/22610

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+                    if (predicate.test(entry)) {
+                        future.complete(position);
+                        return;
+                    }
+                    PositionImpl previousPosition = ledger.getPreviousPosition((PositionImpl) position);
+                    if (!ledger.isValidPosition(previousPosition)) {
+                        future.complete(previousPosition);
+                    } else {
+                        internalAsyncReverseFindPositionOneByOne(ledger, predicate,
+                                ledger.getPreviousPosition((PositionImpl) position), future);
```
Comment (Demogorgon314) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java:67]:
```java
internalAsyncReverseFindPositionOneByOne(ledger, predicate, previousPosition, future);
```
Why do we calculate the `previousPosition` again?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * Reverse find last valid position one-entry by one-entry.
+     */
+    public static CompletableFuture<Position> asyncGetLastValidPosition(final ManagedLedgerImpl ledger,
+                                                                        final Predicate<Entry> predicate,
+                                                                        final PositionImpl startPosition) {
+        CompletableFuture<Position> future = new CompletableFuture<>();
+        if (!ledger.isValidPosition(startPosition)) {
+            future.complete(startPosition);
+        } else {
```
Comment (lhotari) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java:43]: this check could be in the beginning of `internalAsyncReverseFindPositionOneByOne` method to reduce duplication.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+                final Position position = entry.getPosition();
+                try {
+                    if (predicate.test(entry)) {
+                        future.complete(position);
+                        return;
+                    }
+                    PositionImpl previousPosition = ledger.getPreviousPosition((PositionImpl) position);
+                    if (!ledger.isValidPosition(previousPosition)) {
+                        future.complete(previousPosition);
+                    } else {
```
Comment (lhotari) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java:65]: this check could be in the beginning of `internalAsyncReverseFindPositionOneByOne` method to reduce duplication.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java (hunk 16 above)
Comment (dao-jun) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java:43]: fixed by https://github.com/apache/pulsar/pull/22610

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java (hunk 15 above)
Comment (dao-jun) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/ManagedLedgerImplUtils.java:67]: fixed by https://github.com/apache/pulsar/pull/22610
